### ISSUE TITLE
Improve sycing of address with many txs

### ISSUE THREAD

**Issue by chedieck** [PayButton/paybutton-server#370]
Thanks Scott, I was able to reproduce the error here.

The problem is that there are over 700k txs for the second address, so the current implementation goes on to sync them 100 at a time, and it just never finished.

I though about some some possible improvements to this, which I could address in some subsequent PRs (because this issue is  independent of this PR):

1. Progressively commit the tx syncing DB transactions:
   - Transactions would be added, their prices synced, e.g. 1000 at a time (for what I tested there is a byte limit when fetching this data, and for this address it got violated after ~5000 txs at once) and this would already show up in the frontend. It will still take a long time to sync all of them, but we would have the progress feedback.

2. Use another API to get price data
   - The coin.dance API we are using right now only allows to get the price of ONE day a time, which considerably slows the process of getting all txs prices.

3. Have a limit on how many txs to sync, until more is asked
   - Example: Fetch 1000 txs for some address. Only when the user, through the frontend, goes to e.g. page 50 of the tx list and expects to see older transactions do we then fetch the next 1000, and so on.

Any combination of the above is possible, so I am opening another issue  for we to discuss these solutions and also other ideas you guys might have.

_Originally posted by @chedieck in https://github.com/PayButton/paybutton-server/issues/350#issuecomment-1402423529_

**Comment by vorg-code** [comment]
For the second point, we could simply download just once the last 4 years of prices, it is a small dataset to have in a file to initialize the database upon first build, since it will be 4 years * 365 days * 2 coins = less than 3000, it takes a long time now because it is one request per day that we don't have in the database, if it is all in the database, no HTTTP requests will be necessary (unless it is earlier than the 4 years, which must be super rare, we can also increase to 7 years anyway and almost eliminate this api request, leave it just as a safe fallback).

To keep this solution top notch performance-wise, I think it would be nice to update this file from time to time with recent data and also create a job that pre-requests all days missing until today on startup, so that we don't need to do this when the transactions are coming in.

@Klakurka @chedieck what do you think? If you like this solution, I can start working in a PR specifically for this point 2 now and later we discuss points 1 and 3.

**Comment by chedieck** [comment]
> @Klakurka @chedieck what do you think? If you like this solution, I can start working in a PR specifically for this point 2 now and later we discuss points 1 and 3.

I like it, I think we could probably have a `yarn ...` command defined on `package.json` to do that (probably running some other script), the prices file `.gitignore`d, the startup script running the said yarn command when the file doesn't exists to create it (similar to how yarn does with the `node_modules`)

We also don't need more than 4 years because both XEC & BCH have existed for less than that.

**Comment by Klakurka** [comment]
So we'd save this data as part of the initial build and "top it up" with recent prices that we haven't downloaded ahead of time?

**Comment by vorg-code** [comment]
@Klakurka yep.

@chedieck why not leaving it there on the github as well? It would spare time and the data endpoint. It probably could be a bad practice, but I don't see why.

**Comment by chedieck** [comment]
> @Klakurka yep.
> 
> @chedieck why not leaving it there on the github as well? It would spare time and the data endpoint. It probably could be a bad practice, but I don't see why.

I think the problem of leaving it on git is that the file will always be out-of-date, since everyday there is a new price.  So, indeed, at first is less work because e.g. new people cloning the repo already have some prices on the file and only have to fetch the latest.

But then, everytime we e.g. merge master into a branch, we would put the obsolete file there. Or we would have to commit the updated file in every branch, which is kind of dirty. Whereas, by ignoring it, we can develop and always have it updated (by the job you mentioned, for example), without having to stress too much the endpoint or our machine (besides the first time each of the developers run it) and without polluting PRs with it.

**Comment by Klakurka** [comment]
I'm good with either option as long as it works and doesn't make it more difficult to develop/test/etc.

**Comment by vorg-code** [comment]
How about a solution in the middle: I provide a git tracked file with the prices until Feb 1st, we never update that file so no commits will be poluted, each devs enviroment just syncs from this date onwards with the worker thread, also, this worker does not update the file, it only updates the database. I will create a method to populate the file from scratch but it will only be manually run once (we can later run it again to update the file just once a year, if people feel the need to next year, for example).

**Comment by ScottMcDermid** [comment]
> I like it, I think we could probably have a yarn ... command defined on package.json to do that (probably running some other script), the prices file .gitignored, the startup script running the said yarn command when the file doesn't exists to create it (similar to how yarn does with the node_modules)

I like the sound of this approach.

> I provide a git tracked file with the prices until Feb 1st, we never update that file so no commits will be poluted, each devs enviroment just syncs from this date onwards with the worker thread, also, this worker does not update the file, it only updates the database.

I feel like we can get away with sharing our price file with each other if we don't want to sync from the beginning of time. Then we essentially have estavao's solution.

**Comment by chedieck** [comment]
> How about a solution in the middle: I provide a git tracked file with the prices until Feb 1st, we never update that file so no commits will be poluted, each devs enviroment just syncs from this date onwards with the worker thread, also, this worker does not update the file, it only updates the database. I will create a method to populate the file from scratch but it will only be manually run once (we can later run it again to update the file just once a year, if people feel the need to next year, for example).

I don't think this would be very good because then upon every reset of the containers, we would have the worker updating every price since Feb 1st (which is also arbitrary), so e.g. in October that would be ~240 requests to be run if I run `make reset-dev`. If we gitignore the file, we can have the worker updating it everyday it runs, so it would always update at most (today - last day I worked) days, and we would not have to worry about having to keep unstaged changes

**Comment by vorg-code** [comment]
I see now the advantage of keeping the new prices obtained in a file. So we could have 2 files then: 

- one in git with the bulk of prices since the beginning of time until Feb 1st (arbitrary indeed, but so will any date be) to speed up the process of any new installation
- one not tracked by git that the local environment creates when fetching the prices since the first file until today. Code will update the database and this file at the same time when it runs. 

Any command that installs or resets the database will always look for the two files. That seems to be the fastest setup.

**Comment by chedieck** [comment]
I still think is a bad idea to keep a git tracked file.

What you proposed now would bring the advantage of having some prices saved from `beginning` to `Feb 1st 2023`, but this would only make a difference on the first time some new dev clones the repo and spin up the containers for the first time.

Meanwhile, we sacrifice cleanliness by now having to pay attention into reading two files, one that goes from `beginning` to `Feb 1st 2023` and another that goes from `Feb 2nd 2023` to `today`. One that is .gitignored and the other that is not.  

By having only one file, .gitignored, we have the aforementioned disadvantage of having the first time some dev clones the repo taking somewhat longer to sync the files, which is not that big of a hassle, since the requests can all be send in parallel and it should therefore not take much longer than `n_of_requests / api_throttle_rate`. This minor problem is also easily circumvented by @ScottMcDermid's proposal of sharing the file.  And then, after that, every dev will have the file always up-to-date and never have to think about it.

I just don't think it is worth to sacrifice the cleanliness.

**Comment by chedieck** [comment]
> I also feel that it is more of a hassle to wait for the whole sync in the backend since the beggining of history that could take minutes than to have 1 git file tracked.

Even if amounts to minutes, it's only once. Probably would be even faster than building `node_modules` from scratch. 

And it's very much like `node_modules`. It takes some time to build it. But It gets obsolete everyday. You could argue that we should commit `node_modules`, such that new devs don't have to wait several minutes for it to build from zero, but only update the obsolete committed `node_modules`. To make the comparison more accurate, think of a way that this is doable by having `node_modules1` and `node_modules2`: the first one containing the obsolete committed version, the second one containing the updates.

Needless to say this would be a bad practice. There is no need for it, `node_modules` will be built only once. It's also good to have so that I can delete `node_modules` and the whole thing will automatically build it up again.

> imagine a dev spinning it up and trying to sync a wallet in the first minutes, it won't work until the sync is done 

Since this is seeding the database, it should be like `prisma seed` and happen before next is up, so this wouldn't happen because the container would only be up after the database is up & seeded. I am thinking of something like `yarn seedPrices || exit 1`  between these two lines on `scripts/paybutton-server-start.sh`:

![image](https://user-images.githubusercontent.com/21281174/217892972-6a36d202-98cb-426c-b3cc-0da7680e1746.png)

**Comment by vorg-code** [comment]
I see two very important differences between this case and `node_modules`:

1. `node_modules` gets obsolete every day, while this data file does not, past price will be exactly the same forever
2.`node_modules` only gets data from npm and run scripts locally, it doesn't get external third parties apis and neither builds database data

The initial price sync took almost 14 minutes, 148kb, 2840 lines (each line for one pair [date, coin]):

<img width="200" alt="image" src="https://user-images.githubusercontent.com/124014001/217978233-f194fb3f-f72e-4d89-a770-56c937bff449.png">

I still think it is best to have this one file there so that no one needs to get it again, but it is a small matter, whatever decision you guys make I will implement it.

Oh, nice to see that yarn before the container is up, we won't need to handle concurrency of price sync and user action.

**Comment by chedieck** [comment]
Definetively shouldn't take that long. Are you running the requests in parallel or in series?

I made a python script to test how long would take to get a year of prices, it took 2.7s

![image](https://user-images.githubusercontent.com/21281174/218044241-87c2d9d1-77a1-4a6a-8221-9f9999825937.png)

[Here](https://paste.sh/TeAclWFf#Rx8ctrBwdFQ-xTgcyBxG1ZG4) is the code

**Comment by vorg-code** [comment]
PR #402 should solve item 2.
Issue #403 should solve items 1 and 3.
